OpenRA.Mods.Common/Traits/Render/WithSiloSpriteBody.cs

```diff
@@ -0,0 +1,77 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits.Render
+{
+	[Desc("Render trait for buildings that change the sprite according to the remaining resource storage capacity across all depots.")]
+	public class WithSiloSpriteBodyInfo : WithSpriteBodyInfo
+	{
+		[Desc("Internal resource stages. Does not have to match number of sequence frames.")]
+		public readonly int Stages = 10;
+
+		public override object Create(ActorInitializer init) { return new WithSiloSpriteBody(init, this); }
+
+		public override IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer init, RenderSpritesInfo rs, string image, int facings, PaletteReference p)
+		{
+			if (!EnabledByDefault)
+				yield break;
+
+			var anim = new Animation(init.World, image);
+			anim.PlayFetchIndex(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), Sequence), () => 0);
+
+			yield return new SpriteActorPreview(anim, () => WVec.Zero, () => 0, p, rs.Scale);
+		}
+	}
+
+	public class WithSiloSpriteBody : WithSpriteBody, INotifyOwnerChanged
+	{
+		readonly WithSiloSpriteBodyInfo info;
+		PlayerResources playerResources;
+
+		public WithSiloSpriteBody(ActorInitializer init, WithSiloSpriteBodyInfo info)
+			: base(init, info, () => 0)
+		{
+			this.info = info;
+			playerResources = init.Self.Owner.PlayerActor.Trait<PlayerResources>();
+			ConfigureAnimation(init.Self);
+		}
+
+		void ConfigureAnimation(Actor self)
+		{
+			DefaultAnimation.PlayFetchIndex(NormalizeSequence(self, Info.Sequence),
+				() => playerResources.ResourceCapacity != 0
+					? ((info.Stages * DefaultAnimation.CurrentSequence.Length - 1) * playerResources.Resources) / (info.Stages * playerResources.ResourceCapacity)
+					: 0);
+		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			playerResources = newOwner.PlayerActor.Trait<PlayerResources>();
+			ConfigureAnimation(self);
```

Comment:
Why is this needed here?

Reply:
I was assuming this is necessary to refresh on owner change?

Reply:
The animation accesses the resources via the `playerResources` variable, so updating its value is enough.